# Maven Pizza Challenge
In this project I played the role of a BI Consultant hired by **Plato's Pizza**, a Greek-inspired pizza place in New Jersey. I've been hired  to help the restaurant use data to improve operations, and just received the following note:

_Welcome aboard, we're glad you're here to help!_

_Things are going OK here at Plato's, but there's room for improvement. We've been collecting transactional data for the past year, but really haven't been able to put it to good use. Hoping you can analyze the data and put together a report to help us find opportunities to drive more sales and work more efficiently._

_Here are some questions that we'd like to be able to answer:_

* _What days and times do we tend to be busiest?_
* _How many pizzas are we making during peak periods?_
* _What are our best and worst selling pizzas?_
* _What's our average order value?_

_That's all I can think of for now, but if you have any other ideas I'd love to hear them – you're the expert!_

_Thanks in advance,_

_Mario Maven (Manager, Plato's Pizza)_

<center><img src="pizza.jpg"/></center>


## Importing libraries/Modules

In [2]:
import pandas as pd
import datetime as dt
import plotly.offline as po
import plotly.graph_objects as go
from plotly import subplots
import plotly.express as px
import matplotlib.pyplot as plt
import calendar

## Loading Data

In [3]:
ord_det = pd.read_csv('order_details.csv')
ord = pd.read_csv('orders.csv')
piza_typs = pd.read_csv('pizza_types.csv', encoding='cp1252')
piza = pd.read_csv('pizzas.csv')

## Order Table


In [4]:
ord

order_id        date      time
0             1  2015-01-01  11:38:36
1             2  2015-01-01  11:57:40
2             3  2015-01-01  12:12:28
3             4  2015-01-01  12:16:31
4             5  2015-01-01  12:21:30
...         ...         ...       ...
21345     21346  2015-12-31  20:51:07
21346     21347  2015-12-31  21:14:37
21347     21348  2015-12-31  21:23:10
21348     21349  2015-12-31  22:09:54
21349     21350  2015-12-31  23:02:05

[21350 rows x 3 columns]

In [5]:
ord.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21350 entries, 0 to 21349
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   order_id  21350 non-null  int64 
 1   date      21350 non-null  object
 2   time      21350 non-null  object
dtypes: int64(1), object(2)
memory usage: 500.5+ KB


In [6]:
ord[ord.duplicated()]

Empty DataFrame
Columns: [order_id, date, time]
Index: []

## Order Detatils Table


In [7]:
ord_det

order_details_id  order_id       pizza_id  quantity
0                     1         1     hawaiian_m         1
1                     2         2  classic_dlx_m         1
2                     3         2  five_cheese_l         1
3                     4         2    ital_supr_l         1
4                     5         2     mexicana_m         1
...                 ...       ...            ...       ...
48615             48616     21348  ckn_alfredo_m         1
48616             48617     21348  four_cheese_l         1
48617             48618     21348   napolitana_s         1
48618             48619     21349     mexicana_l         1
48619             48620     21350      bbq_ckn_s         1

[48620 rows x 4 columns]

In [8]:
ord_det.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48620 entries, 0 to 48619
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   order_details_id  48620 non-null  int64 
 1   order_id          48620 non-null  int64 
 2   pizza_id          48620 non-null  object
 3   quantity          48620 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB


In [9]:
ord_det[ord_det.duplicated()]

Empty DataFrame
Columns: [order_details_id, order_id, pizza_id, quantity]
Index: []

## Pizzas Table

In [10]:
piza

pizza_id pizza_type_id size  price
0       bbq_ckn_s       bbq_ckn    S  12.75
1       bbq_ckn_m       bbq_ckn    M  16.75
2       bbq_ckn_l       bbq_ckn    L  20.75
3      cali_ckn_s      cali_ckn    S  12.75
4      cali_ckn_m      cali_ckn    M  16.75
..            ...           ...  ...    ...
91  spinach_fet_m   spinach_fet    M  16.00
92  spinach_fet_l   spinach_fet    L  20.25
93   veggie_veg_s    veggie_veg    S  12.00
94   veggie_veg_m    veggie_veg    M  16.00
95   veggie_veg_l    veggie_veg    L  20.25

[96 rows x 4 columns]

In [11]:
piza.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96 entries, 0 to 95
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   pizza_id       96 non-null     object 
 1   pizza_type_id  96 non-null     object 
 2   size           96 non-null     object 
 3   price          96 non-null     float64
dtypes: float64(1), object(3)
memory usage: 3.1+ KB


In [12]:
piza[piza.duplicated()]

Empty DataFrame
Columns: [pizza_id, pizza_type_id, size, price]
Index: []

## PIzza Types Table


In [13]:
piza_typs.head()

pizza_type_id                          name category  \
0       bbq_ckn    The Barbecue Chicken Pizza  Chicken   
1      cali_ckn  The California Chicken Pizza  Chicken   
2   ckn_alfredo     The Chicken Alfredo Pizza  Chicken   
3     ckn_pesto       The Chicken Pesto Pizza  Chicken   
4    southw_ckn   The Southwest Chicken Pizza  Chicken   

                                         ingredients  
0  Barbecued Chicken, Red Peppers, Green Peppers,...  
1  Chicken, Artichoke, Spinach, Garlic, Jalapeno ...  
2  Chicken, Red Onions, Red Peppers, Mushrooms, A...  
3  Chicken, Tomatoes, Red Peppers, Spinach, Garli...  
4  Chicken, Tomatoes, Red Peppers, Red Onions, Ja...

In [14]:
piza_typs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   pizza_type_id  32 non-null     object
 1   name           32 non-null     object
 2   category       32 non-null     object
 3   ingredients    32 non-null     object
dtypes: object(4)
memory usage: 1.1+ KB


In [15]:
piza_typs[piza_typs.duplicated()]

Empty DataFrame
Columns: [pizza_type_id, name, category, ingredients]
Index: []

## Merging The Four tables

After going through the data, i found no empty row neither di i find any duplicates. Looking at The metadata, there are unique identifiers and foreign keys that can help maerge these tables.

I will be merging the tables below.

In [16]:
pizza_order = ord.merge(right=ord_det, on='order_id', how='left').\
                    merge(right=piza, on='pizza_id', how='left').\
                    merge(right=piza_typs, on='pizza_type_id', how='left')

pizza_order.head(3)

order_id        date      time  order_details_id       pizza_id  quantity  \
0         1  2015-01-01  11:38:36                 1     hawaiian_m         1   
1         2  2015-01-01  11:57:40                 2  classic_dlx_m         1   
2         2  2015-01-01  11:57:40                 3  five_cheese_l         1   

  pizza_type_id size  price                      name category  \
0      hawaiian    M  13.25        The Hawaiian Pizza  Classic   
1   classic_dlx    M  16.00  The Classic Deluxe Pizza  Classic   
2   five_cheese    L  18.50     The Five Cheese Pizza   Veggie   

                                         ingredients  
0           Sliced Ham, Pineapple, Mozzarella Cheese  
1  Pepperoni, Mushrooms, Red Onions, Red Peppers,...  
2  Mozzarella Cheese, Provolone Cheese, Smoked Go...

I combined the date and time column to have a datetime column. Also I calculate the total for each order.

In [17]:
total = pd.Series(pizza_order['quantity'] * pizza_order['price'])
pizza_order.insert(9, 'total', total)
date_time = pd.Series(pizza_order['date'] + " " + pizza_order['time'])
date_time = pd.to_datetime(date_time)
pizza_order.insert(3, 'date_time', date_time)
pizza_order.drop(['date', 'time'],axis=1,inplace=True)
pizza_order.set_index('date_time', inplace=True)

In [18]:
pizza_order.head(3)

order_id  order_details_id       pizza_id  quantity  \
date_time                                                                  
2015-01-01 11:38:36         1                 1     hawaiian_m         1   
2015-01-01 11:57:40         2                 2  classic_dlx_m         1   
2015-01-01 11:57:40         2                 3  five_cheese_l         1   

                    pizza_type_id size  price  total  \
date_time                                              
2015-01-01 11:38:36      hawaiian    M  13.25  13.25   
2015-01-01 11:57:40   classic_dlx    M  16.00  16.00   
2015-01-01 11:57:40   five_cheese    L  18.50  18.50   

                                         name category  \
date_time                                                
2015-01-01 11:38:36        The Hawaiian Pizza  Classic   
2015-01-01 11:57:40  The Classic Deluxe Pizza  Classic   
2015-01-01 11:57:40     The Five Cheese Pizza   Veggie   

                                                           ingredients  
date_time                                                               
2015-01-01 11:38:36           Sliced Ham, Pineapple, Mozzarella Cheese  
2015-01-01 11:57:40  Pepperoni, Mushrooms, Red Onions, Red Peppers,...  
2015-01-01 11:57:40  Mozzarella Cheese, Provolone Cheese, Smoked Go...

In [128]:
pizza_order.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 48620 entries, 2015-01-01 11:38:36 to 2015-12-31 23:02:05
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          48620 non-null  int64  
 1   order_details_id  48620 non-null  int64  
 2   pizza_id          48620 non-null  object 
 3   quantity          48620 non-null  int64  
 4   pizza_type_id     48620 non-null  object 
 5   size              48620 non-null  object 
 6   price             48620 non-null  float64
 7   total             48620 non-null  float64
 8   name              48620 non-null  object 
 9   category          48620 non-null  object 
 10  ingredients       48620 non-null  object 
dtypes: float64(2), int64(3), object(6)
memory usage: 4.5+ MB


## Exploring the data for insight.

### Some questions that we'd like to answer:
* What days and times do we tend to be busiest?
* How many pizzas are we making during peak periods?
* What are our best and worst selling pizzas?
* What's our average order value?


### Partitioning the day into different section
* Early Morning: 6 to 9 am 
* Late Morning: 9 am to 12pm
* early afternoon: 12 to 3 pm
* Late afternoon 3 to 6pm
* evening: 6 pm to 9 pm
* late evening: 9pm to 00:00
* late at night: 00:00 to 6 am

There is a large amount of order between the hour of 12:00 p.m and 15:00 pm (Early Afternoon)

In [148]:
early_morning = len(pizza_order.between_time('06:00', '09:00'))
late_morning = len(pizza_order.between_time('09:00', '12:00'))
early_afternoon = len(pizza_order.between_time('12:00', '15:00'))
late_afternoon = len(pizza_order.between_time('15:00', '18:00'))
evening = len(pizza_order.between_time('18:00', '21:00'))
late_evening = len(pizza_order.between_time('21:00', '00:00'))
late_at_night = len(pizza_order.between_time('00:00', '06:00'))

dic = {
    'early_morning': ['6 a.m - 9 a.m', early_morning],
    'late_morning': ['9 a.m - 12 p.m', late_morning],
    'early_afternoon': ['12 p.m - 3 p.m', early_afternoon],
    'late_afternoon': ['3 p.m - 6 p.m', late_afternoon],
    'evening': ['6 p.m - 9 p.m', evening],
    'late_evening': ['9 p.m - 12 a.m', late_evening],
    'late_at_night': ['12 a.m - 6 a.m', late_at_night]
}

df = pd.DataFrame.from_dict(dic, orient='index').reset_index()
df.rename(columns={'index':'day_period', 0:'period', 1:'order_count'}, inplace=True)


fig = px.line(df, x='day_period', y='order_count', 
              markers=True, text='order_count', width=900,
              title='Total pizza order within the day',
              labels={'day_period':'Day Period', 'order_count':'Order Count'})
fig.update_traces(textposition='top right',texttemplate="%{text:.2s}")

In [142]:
jan = pizza_order.loc[(pizza_order.index > '2015-01-01 00:00:00') & (pizza_order.index <= '2015-01-31 00:00:00')]
feb = pizza_order.loc[(pizza_order.index > '2015-02-01 00:00:00') & (pizza_order.index <= '2015-02-28 00:00:00')]

def pizza_count_peak_period(dataframe, title):
    early_morning = len(dataframe.between_time('06:00', '09:00'))
    late_morning = len(dataframe.between_time('09:00', '12:00'))
    early_afternoon = len(dataframe.between_time('12:00', '15:00'))
    late_afternoon = len(dataframe.between_time('15:00', '18:00'))
    evening = len(dataframe.between_time('18:00', '21:00'))
    late_evening = len(dataframe.between_time('21:00', '00:00'))
    late_at_night = len(dataframe.between_time('00:00', '06:00'))

    dic = {
        'early_morning': ['6 a.m - 9 a.m', early_morning],
        'late_morning': ['9 a.m - 12 p.m', late_morning],
        'early_afternoon': ['12 p.m - 3 p.m', early_afternoon],
        'late_afternoon': ['3 p.m - 6 p.m', late_afternoon],
        'evening': ['6 p.m - 9 p.m', evening],
        'late_evening': ['9 p.m - 12 a.m', late_evening],
        'late_at_night': ['12 a.m - 6 a.m', late_at_night]
    }

    df = pd.DataFrame.from_dict(dic, orient='index').reset_index()
    df.rename(columns={'index':'day_period', 0:'period', 1:'order_count'}, inplace=True)


    fig = px.line(df, x='day_period', y='order_count', 
                markers=True, text='order_count', width=900,
                title=title,
                labels={'day_period':'Day Period', 'order_count':'Order Count'})
    return fig.update_traces(textposition='top right',texttemplate="%{text:.2s}")

Comparing pizzas made during peak period for two months. 

In [145]:
pizza_count_peak_period(jan, 'January order within the day')

In [147]:
pizza_count_peak_period(feb, 'Febuary order within the day')

The average quantity of pizza ordered is very high around 10 a.m and reduces as the hour goes by.

In [84]:
a = pizza_order.groupby(by=pizza_order.index.hour)['quantity'].mean()
trace = px.line(a,
    x=a.index,
    y=a.values,
    orientation='h',
    width=900, 
    labels={'date_time':'Hour', 'y':'Order Counts'},
    title='Average quantity by the hour'
)
trace.update_traces(texttemplate="%{text:.2s}", textposition='top center')

Fridays turn out to be the most busiest day as the pizza ordered during this day is at the peak.

In [88]:
days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']
a = pizza_order.groupby(by=pizza_order.index.day_name())['order_id'].count().reindex(days)
trace = px.line(a,
    x=a.index,
    y=a.values,
    markers=True,
    text=a.values, width=900,
    labels={'date_time':'Weekday', 'y':'Order Counts'},
    title='Pizza order by weekday'
)
trace.update_traces(texttemplate="%{text:.2s}", textposition='top left')

Average sales is high on Tuesdays and Thursdays.

In [164]:
days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']
a = pizza_order.groupby(by=pizza_order.index.day_name())['total'].mean().reindex(days)
trace = px.line(a,
    x=a.index,
    y=a.values,
    markers=True,
    text=a.values, width=900,
    labels={'date_time':'Weekday', 'y':'Order Counts'},
    title='Average sales by weekday'
)
trace.update_traces(texttemplate="%{text:.4s}", textposition='top left')

Average order value

In [149]:
pizza_order.total.mean()

16.821473673385437

The best selling pizza is The Classic Deluxe while the worst selling pizza is The Soppressatan Pizza

In [25]:
def simple_bar(data, color):
   bar = go.Bar(
       x=data.index,
       y=data.values,
       marker=dict(color=color),
       text=data.values,
       orientation='v',
   )
   return bar

def double_bar_count(color1, color2, title1, title2, title3):
    most = pizza_order.groupby('name')['price'].count().sort_values(ascending=False)[:5]
    trace1 = simple_bar(most, color1)
    least = pizza_order.groupby('name')['price'].count().sort_values(ascending=False)[-6:-1]
    trace2 = simple_bar(least, color2) 

    fig1 = subplots.make_subplots(rows=1, cols=2, subplot_titles=(title1, title2))
    fig1.append_trace(trace1, 1,1)
    fig1.append_trace(trace2, 1,2)
    fig1.update_traces(texttemplate="%{text:,.0f}")
    fig1['layout'].update(height=500, title=title3, showlegend=False, width=1000)
    return po.iplot(fig1)

def double_bar_sum(color1, color2, title1, title2, title3):
    most = pizza_order.groupby('name')['total'].sum().sort_values(ascending=False)[:5]
    trace1 = simple_bar(most, color1)
    least = pizza_order.groupby('name')['total'].sum().sort_values(ascending=False)[-6:-1]
    trace2 = simple_bar(least, color2) 

    fig1 = subplots.make_subplots(rows=1, cols=2, subplot_titles=(title1, title2))
    fig1.append_trace(trace1, 1,1)
    fig1.append_trace(trace2, 1,2)
    fig1.update_traces(texttemplate="%{text:,.0f}")
    fig1['layout'].update(height=500, title=title3, showlegend=False, width=1000)
    return po.iplot(fig1)

In [26]:
double_bar_count('orange', 'navy', '5 Most Ordered Pizza', '5 Least Ordered Pizza', 'Pizza Orders')

In [27]:
double_bar_sum('gold', 'tomato', '5 Most Sold Pizza', '5 Least Sold Pizza', 'Pizza Sales')

The highest sale was made in July with average of $4,301, while the least sale was made in October with average of $3,797.

In [167]:
monthly_order = pizza_order.groupby(by=pizza_order.index.month)['order_id'].count()
monthly_order =  monthly_order.reset_index(name='count')
monthly_order['date_time'] = monthly_order['date_time'].apply(lambda x: calendar.month_abbr[x])

trace = px.line(monthly_order,
    x='date_time',
    y='count',
    markers=True,
    text='count', width=900,
    # labels={'date_time':'Weekday', 'y':'Order Counts'},
    title='Average sales by weekday'
)
trace.update_traces(texttemplate="%{text:.3s}", textposition='top left').show()


monthly_sales = pizza_order.groupby(by=pizza_order.index.month)['total'].sum()
monthly_sales =  monthly_sales.reset_index(name='count')
monthly_sales['date_time'] = monthly_sales['date_time'].apply(lambda x: calendar.month_abbr[x])

trace = px.bar(monthly_sales,
    x='date_time',
    y='count',
    text_auto='.3s',
    title='Monthly Sales',
    width=900,
    labels={'date_time':'Month', 'count':'Sales'},
)
trace.update_traces(textposition='outside', cliponaxis=False)
trace.show()

Small size pizza from the veggie category is the most purchase pizza.

In [28]:
def cat_count(col1, text, col2):
    fil = pizza_order[pizza_order[col1] == text]
    counts = fil[col2].value_counts().reset_index()
    return counts

def traces(data, name, dom_x_y, dom1, dom2):
    trace = go.Pie(
    labels=data['index'],
    values=data['size'],
    name=name,
    hole=.5,
    domain={dom_x_y: [dom1, dom2]},
    direction='clockwise',
    marker=dict(colors=['red', 'navy', 'cyan', 'orange', 'lightgrey']))
    return trace


count_clas = cat_count('category', 'Classic', 'size')
count_sup = cat_count('category', 'Supreme', 'size')
count_veg = cat_count('category', 'Veggie', 'size')
count_chi = cat_count('category', 'Chicken', 'size')

trace1 = traces(count_clas, 'classic', 'x', 0, .30)
trace2 = traces(count_sup, 'supreme', 'x', .7, 1)
trace3 = traces(count_veg, 'veggie', 'y', 0, .48)
trace4 = traces(count_chi, 'chicken', 'y', .52, 1)


layout = dict(
    title='Pizza Order by Category Size',
    font=dict(size=12),
    height=560, 
    legend=dict(orientation='h'),
    width=700,
    annotations= [
        dict(
            x=.1, y=.5,
            text='Classic',
            showarrow=False,
            font=dict(size=15)
        ),
        dict(
            x=.92, y=.5,
            text='Supreme',
            showarrow=False,
            font=dict(size=15)
        ),
        dict(
            x=.5, y=.2,
            text='Veggie',
            showarrow=False,
            font=dict(size=15)
        ),
        dict(
            x=.5, y=.8,
            text='Chicken',
            showarrow=False,
            font=dict(size=13)
        )
    ]
    
)


fig = dict(data=[trace1, trace2, trace3, trace4], layout=layout)
po.iplot(fig)

Sunburst of each pizza category, displaying the sale made from each size.

In [93]:
clas = pizza_order[pizza_order['category'] == 'Classic']
fig = px.sunburst(clas, path=['category', 'name', 'size'], values='total', width=500, 
                  title='Sales by size for each pizza type in classic category')
fig.show()

sup = pizza_order[pizza_order['category'] == 'Supreme']
fig = px.sunburst(sup, path=['category', 'name', 'size'], values='total', width=500,
                  title='Sales by size for each pizza type in supreme category')
fig.show()

chi = pizza_order[pizza_order['category'] == 'Chicken']
fig = px.sunburst(chi, path=['category', 'name', 'size'], values='total', width=500,
                  title='Sales by size for each pizza type in chicken category')
fig.show()

veg = pizza_order[pizza_order['category'] == 'Veggie']
fig = px.sunburst(veg, path=['category', 'name', 'size'], values='total', width=500,
                  title='Sales by size for each pizza type in veggie category')
fig.show()

# Maven Pizza Challenge
In this project I played the role of a BI Consultant hired by **Plato's Pizza**, a Greek-inspired pizza place in New Jersey. I've been hired  to help the restaurant use data to improve operations.

<center><img src="pizza.jpg"/></center>

### Some questions that we'd like to be able to answer:
* What days and times do we tend to be busiest?
* How many pizzas are we making during peak periods?
* What are our best and worst selling pizzas?
* What's our average order value?

## About the dataset
* This dataset contains 4 tables in CSV format
* The Orders table contains the date & time that all 5,238 table orders were placed
* The Order Details table contains the different pizzas served with each order in the Orders table, and their quantities
* The Pizzas table contains the size and price for each distinct pizza in the Order Details table, as well as its broader pizza type
* The Pizza Types table contains details on the pizza types in the Pizzas table, including their name as it appears on the menu, the category it falls under, and its list of ingredients

<center><img src="weekday_order.png"/></center>

## Answers
* The most **busiest day** is _**Friday**_ with order count of **8,106**. 
* More than **1000 pizza are made during peak peiods**. For _January_ record, _1,410 pizza_ was made in the period of _12 pm - 3pm_ while in _February_, _1,275 pizza_ was ordered during the same period _(early afternoon)_
* The **3 best selling pizzas** include, _**The Classic Deluxe**_, _**The Barbecue Chicken Pizza**_, and _**The Hawaiian Pizza**_.
* The **3 worst selling pizza** include, _**The Soppressatan Pizza**_, _**The Spinach Pesto Pizza**_ and _**The Spinach Supreme Pizza**_.
The average order value is **$16.8**.

<center><img src="cat_size_sale.png"/></center>

## Summary
* The average quantity of _pizza ordered is very high around 10 a.m_ and reduces as the hour goes by.
* Average sales is high on _Tuesdays_ and _Thursdays_.
* The _highest sale was made in July with average of $4,301_, while the _least sale was made in October with average of $3,797_.
* Small size pizza from the veggie category is the most purchase pizza.